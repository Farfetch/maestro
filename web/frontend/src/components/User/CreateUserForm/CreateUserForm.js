import { Form, Input, Select } from "antd";
import PropTypes from "prop-types";
import React from "react";

import { userRole as userRoleModel } from "../../../lib/api/models";

const CreateUserForm = ({
  form,
  initialValues,
  onFinish,
  workspaces,
  users
}) => {
  const onFinishFailed = async () => {};

  return (
    <Form
      layout="vertical"
      form={form}
      name="basic"
      onFinish={onFinish}
      onFinishFailed={onFinishFailed}
      initialValues={initialValues}
    >
      <Form.Item label="Name">
        <Form.Item
          name="name"
          noStyle
          rules={[
            {
              required: true,
              message: "Please enter User Name"
            }
          ]}
        >
          <Input placeholder="Enter prefered workspace name" />
        </Form.Item>
      </Form.Item>
      <Form.Item label="Email">
        <Form.Item
          name="email"
          noStyle
          rules={[
            {
              required: true,
              message: "Please enter User Email"
            },
            {
              validator: async (rule, value) => {
                // Skip validation when email didn't change
                if (initialValues && initialValues.email === value) return;
                const user = users.find(({ email }) => email === value);
                if (user) {
                  throw new Error("User already exists");
                }
              }
            }
          ]}
        >
          <Input placeholder="Enter prefered workspace name" />
        </Form.Item>
      </Form.Item>
      <Form.Item label="Role">
        <Form.Item
          name="role"
          noStyle
          rules={[
            {
              required: true,
              message: "Please select at leat one user"
            }
          ]}
        >
          <Select
            mode="single"
            style={{ width: "100%" }}
            placeholder="Select User Role"
          >
            <Select.Option key={userRoleModel.USER}>
              {userRoleModel.USER}
            </Select.Option>
            <Select.Option key={userRoleModel.ADMIN}>
              {userRoleModel.ADMIN}
            </Select.Option>
          </Select>
        </Form.Item>
      </Form.Item>
      <Form.Item label="Workspaces">
        <Form.Item
          name="workspaceIds"
          noStyle
          rules={[
            {
              required: true,
              message: "Please select at least one workspace"
            }
          ]}
        >
          <Select
            mode="multiple"
            allowClear
            style={{ width: "100%" }}
            placeholder="Select available workspaces"
          >
            {workspaces.map((workspace) => (
              <Select.Option key={workspace.id}>{workspace.name}</Select.Option>
            ))}
          </Select>
        </Form.Item>
      </Form.Item>
    </Form>
  );
};

CreateUserForm.propTypes = {
  form: PropTypes.any.isRequired,
  initialValues: PropTypes.shape({
    name: PropTypes.string,
    workspaceIds: PropTypes.arrayOf(PropTypes.string)
  }),
  onFinish: PropTypes.func.isRequired,
  workspaces: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired
    })
  ),
  users: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      email: PropTypes.string.isRequired,
      role: PropTypes.oneOf([userRoleModel.USER, userRoleModel.ADMIN]),
      workspaceIds: PropTypes.arrayOf(PropTypes.string),
      createdAt: PropTypes.object.isRequired,
      updatedAt: PropTypes.object.isRequired
    })
  )
};

export default CreateUserForm;
