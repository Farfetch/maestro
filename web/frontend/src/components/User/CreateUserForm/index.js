import CreateUserForm from "./CreateUserForm";

export default CreateUserForm;
