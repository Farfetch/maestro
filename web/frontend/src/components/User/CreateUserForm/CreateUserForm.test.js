import { cleanup, render, waitFor } from "@testing-library/react";
import moment from "moment";
import React from "react";

import CreateUserForm from "./CreateUserForm";

afterEach(cleanup);

describe("components/User/CreateUserForm", () => {
  test(`should render CreateUserForm component`, async () => {
    const users = [
      {
        id: "user-id-1",
        name: "user 1",
        email: "user1@maestro.net",
        role: "USER",
        workspaceIds: ["1-2-3", "4-5-6"],
        createdAt: moment("2018-05-24 13:48:00"),
        updatedAt: moment("2018-05-24 13:48:00")
      },
      {
        id: "user-id-2",
        name: "user 2",
        email: "user2@maestro.net",
        role: "USER",
        workspaceIds: ["4-5-6"],
        createdAt: moment("2018-05-24 13:48:00"),
        updatedAt: moment("2018-05-24 13:48:00")
      }
    ];

    const workspaces = [
      {
        id: "1-2-3",
        name: "Workspace 1",
        createdAt: moment("2018-05-24 13:48:00"),
        updatedAt: moment("2018-05-24 13:48:00")
      },
      {
        id: "4-5-6",
        name: "Workspace 2",
        createdAt: moment("2018-05-24 13:48:00"),
        updatedAt: moment("2018-05-24 13:48:00")
      }
    ];

    const { container } = render(
      <CreateUserForm
        form={null}
        initialValues={{}}
        onFinish={() => null}
        users={users}
        workspaces={workspaces}
      />
    );

    await waitFor(() => {
      expect(container.outerHTML).toMatchSnapshot();
    });
  });

  test(`should render CreateUserForm component with Initial values`, async () => {
    const users = [
      {
        id: "user-id-1",
        name: "user 1",
        email: "user1@maestro.net",
        role: "USER",
        workspaceIds: ["1-2-3", "4-5-6"],
        createdAt: moment("2018-05-24 13:48:00"),
        updatedAt: moment("2018-05-24 13:48:00")
      },
      {
        id: "user-id-2",
        name: "user 2",
        email: "user2@maestro.net",
        role: "USER",
        workspaceIds: ["4-5-6"],
        createdAt: moment("2018-05-24 13:48:00"),
        updatedAt: moment("2018-05-24 13:48:00")
      }
    ];

    const workspaces = [
      {
        id: "1-2-3",
        name: "Workspace 1",
        createdAt: moment("2018-05-24 13:48:00"),
        updatedAt: moment("2018-05-24 13:48:00")
      },
      {
        id: "4-5-6",
        name: "Workspace 2",
        createdAt: moment("2018-05-24 13:48:00"),
        updatedAt: moment("2018-05-24 13:48:00")
      }
    ];

    const { container } = render(
      <CreateUserForm
        form={null}
        initialValues={users[0]}
        onFinish={() => null}
        users={users}
        workspaces={workspaces}
      />
    );

    await waitFor(() => {
      expect(container.outerHTML).toMatchSnapshot();
    });
  });
});
