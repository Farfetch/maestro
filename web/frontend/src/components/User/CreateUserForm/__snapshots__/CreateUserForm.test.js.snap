// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/User/CreateUserForm should render CreateUserForm component 1`] = `"<div><form id=\\"basic\\" class=\\"ant-form ant-form-vertical\\"><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-col ant-form-item-label\\"><label class=\\"\\" title=\\"Name\\">Name</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><input placeholder=\\"Enter prefered workspace name\\" id=\\"basic_name\\" class=\\"ant-input\\" type=\\"text\\" value=\\"\\"></div></div></div></div><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-col ant-form-item-label\\"><label class=\\"\\" title=\\"Email\\">Email</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><input placeholder=\\"Enter prefered workspace name\\" id=\\"basic_email\\" class=\\"ant-input\\" type=\\"text\\" value=\\"\\"></div></div></div></div><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-col ant-form-item-label\\"><label class=\\"\\" title=\\"Role\\">Role</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><div class=\\"ant-select ant-select-single ant-select-show-arrow\\" style=\\"width: 100%;\\"><div class=\\"ant-select-selector\\"><span class=\\"ant-select-selection-search\\"><input type=\\"search\\" id=\\"basic_role\\" autocomplete=\\"off\\" class=\\"ant-select-selection-search-input\\" role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"basic_role_list\\" aria-autocomplete=\\"list\\" aria-controls=\\"basic_role_list\\" aria-activedescendant=\\"basic_role_list_0\\" readonly=\\"\\" unselectable=\\"on\\" style=\\"opacity: 0;\\" value=\\"\\"></span><span class=\\"ant-select-selection-placeholder\\">Select User Role</span></div><span class=\\"ant-select-arrow\\" style=\\"user-select: none;\\" unselectable=\\"on\\" aria-hidden=\\"true\\"><span role=\\"img\\" aria-label=\\"down\\" class=\\"anticon anticon-down ant-select-suffix\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></span></span></div></div></div></div></div><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-col ant-form-item-label\\"><label class=\\"\\" title=\\"Workspaces\\">Workspaces</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><div class=\\"ant-select ant-select-multiple ant-select-allow-clear ant-select-show-search\\" style=\\"width: 100%;\\"><div class=\\"ant-select-selector\\"><div class=\\"ant-select-selection-overflow\\"><div class=\\"ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix\\" style=\\"opacity: 1;\\"><div class=\\"ant-select-selection-search\\" style=\\"width: 0px;\\"><input type=\\"search\\" id=\\"basic_workspaceIds\\" autocomplete=\\"off\\" class=\\"ant-select-selection-search-input\\" role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"basic_workspaceIds_list\\" aria-autocomplete=\\"list\\" aria-controls=\\"basic_workspaceIds_list\\" aria-activedescendant=\\"basic_workspaceIds_list_0\\" readonly=\\"\\" unselectable=\\"on\\" style=\\"opacity: 0;\\" value=\\"\\"><span class=\\"ant-select-selection-search-mirror\\" aria-hidden=\\"true\\">&nbsp;</span></div></div></div><span class=\\"ant-select-selection-placeholder\\">Select available workspaces</span></div></div></div></div></div></div></form></div>"`;

exports[`components/User/CreateUserForm should render CreateUserForm component with Initial values 1`] = `"<div><form id=\\"basic\\" class=\\"ant-form ant-form-vertical\\"><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-col ant-form-item-label\\"><label class=\\"\\" title=\\"Name\\">Name</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><input placeholder=\\"Enter prefered workspace name\\" id=\\"basic_name\\" class=\\"ant-input\\" type=\\"text\\" value=\\"user 1\\"></div></div></div></div><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-col ant-form-item-label\\"><label class=\\"\\" title=\\"Email\\">Email</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><input placeholder=\\"Enter prefered workspace name\\" id=\\"basic_email\\" class=\\"ant-input\\" type=\\"text\\" value=\\"user1@maestro.net\\"></div></div></div></div><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-col ant-form-item-label\\"><label class=\\"\\" title=\\"Role\\">Role</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><div class=\\"ant-select ant-select-single ant-select-show-arrow\\" style=\\"width: 100%;\\"><div class=\\"ant-select-selector\\"><span class=\\"ant-select-selection-search\\"><input type=\\"search\\" id=\\"basic_role\\" autocomplete=\\"off\\" class=\\"ant-select-selection-search-input\\" role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"basic_role_list\\" aria-autocomplete=\\"list\\" aria-controls=\\"basic_role_list\\" aria-activedescendant=\\"basic_role_list_0\\" readonly=\\"\\" unselectable=\\"on\\" style=\\"opacity: 0;\\" value=\\"\\"></span><span class=\\"ant-select-selection-item\\" title=\\"USER\\">USER</span></div><span class=\\"ant-select-arrow\\" style=\\"user-select: none;\\" unselectable=\\"on\\" aria-hidden=\\"true\\"><span role=\\"img\\" aria-label=\\"down\\" class=\\"anticon anticon-down ant-select-suffix\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></span></span></div></div></div></div></div><div class=\\"ant-row ant-form-item\\"><div class=\\"ant-col ant-form-item-label\\"><label class=\\"\\" title=\\"Workspaces\\">Workspaces</label></div><div class=\\"ant-col ant-form-item-control\\"><div class=\\"ant-form-item-control-input\\"><div class=\\"ant-form-item-control-input-content\\"><div class=\\"ant-select ant-select-multiple ant-select-allow-clear ant-select-show-search\\" style=\\"width: 100%;\\"><div class=\\"ant-select-selector\\"><div class=\\"ant-select-selection-overflow\\"><div class=\\"ant-select-selection-overflow-item\\" style=\\"opacity: 1;\\"><span class=\\"ant-select-selection-item\\" title=\\"Workspace 1\\"><span class=\\"ant-select-selection-item-content\\">Workspace 1</span><span class=\\"ant-select-selection-item-remove\\" style=\\"user-select: none;\\" unselectable=\\"on\\" aria-hidden=\\"true\\"><span role=\\"img\\" aria-label=\\"close\\" class=\\"anticon anticon-close\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"close\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z\\"></path></svg></span></span></span></div><div class=\\"ant-select-selection-overflow-item\\" style=\\"opacity: 1;\\"><span class=\\"ant-select-selection-item\\" title=\\"Workspace 2\\"><span class=\\"ant-select-selection-item-content\\">Workspace 2</span><span class=\\"ant-select-selection-item-remove\\" style=\\"user-select: none;\\" unselectable=\\"on\\" aria-hidden=\\"true\\"><span role=\\"img\\" aria-label=\\"close\\" class=\\"anticon anticon-close\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"close\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z\\"></path></svg></span></span></span></div><div class=\\"ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix\\" style=\\"opacity: 1;\\"><div class=\\"ant-select-selection-search\\" style=\\"width: 0px;\\"><input type=\\"search\\" id=\\"basic_workspaceIds\\" autocomplete=\\"off\\" class=\\"ant-select-selection-search-input\\" role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"basic_workspaceIds_list\\" aria-autocomplete=\\"list\\" aria-controls=\\"basic_workspaceIds_list\\" aria-activedescendant=\\"basic_workspaceIds_list_0\\" readonly=\\"\\" unselectable=\\"on\\" style=\\"opacity: 0;\\" value=\\"\\"><span class=\\"ant-select-selection-search-mirror\\" aria-hidden=\\"true\\">&nbsp;</span></div></div></div></div><span class=\\"ant-select-clear\\" style=\\"user-select: none;\\" unselectable=\\"on\\" aria-hidden=\\"true\\"><span role=\\"img\\" aria-label=\\"close-circle\\" class=\\"anticon anticon-close-circle\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"close-circle\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z\\"></path></svg></span></span></div></div></div></div></div></form></div>"`;
